src/mds/SimpleLock.h

```diff
@@ -211,16 +211,16 @@ class SimpleLock {
     parent->encode_lock_state(type->type, bl);
   }
   void finish_waiters(uint64_t mask, int r=0) {
-    parent->finish_waiting(mask << get_wait_shift(), r);
+    parent->finish_waiting(mask << wait_shift, r);
```

Comment:
I don't think this is correct. For example:

https://github.com/ceph/ceph/blob/8ffdab763374c29194121bdc9c4f48e10cae0da1/src/mds/Locker.cc#L2012

The `Locker` is waking up waiters for `SimpleLock::WAIT_STABLE|SimpleLock::WAIT_WR|SimpleLock::WAIT_RD` but only for that particular `lock`. Not all locks for that `CInode`.

With this change, waiters for any lock of the inode will be awoken if the wait mask fits the current lock state. That behavior did not change with

https://github.com/ceph/ceph/pull/8965/files#diff-d14791b3f532382cab6bf71ecff722c2eb5f5febf4b83ce62190a8e31084212dR1590-R1597

It simply imposed an order on the context completions.

What I think you can do instead is convert `MDSCacheObject::waiting` to:

```
using waitmask = std::bitset<128>;
mempool::mds_co::compact_multimap<waitmask, std::pair<uint64_t, MDSContext*>> waiting;
```

Then for all `SimpleLock` waiter methods, create a `MDSCacheObject::waitmask` which shifts the wait bits left 64 bits (for orderliness not compactness) instead of by `8`. Then call a new overloaded method like `MDSCacheObject::add_waiter(waitmask, MDSContext*)`. The current `MDSCacheObject::add_waiter(uint64_t mask, MDSContext*c)` will simply call `add_waiter(waitmask(mask), c);`

What do you think?

(Good job finding the issue!)

Reply:
Great catch!
I missed that, but now I also realize that we shouldn't need more than a single bit per lock, since the ordering is done with the sequence

Reply:
I don't think so... if we do

`lock->finish_waiters(SimpleLock::WAIT_RD);`

then we do not want to wake up writers (in fact, that could conceivably trigger undesirable state transitions). Right?

Reply:
Now that I had the time to think about it, I did find a single bit per lock solution. The trick is to add 4 bits for the operation type, and then a bit per lock. See my latest change - please review it thoroughly cause I might still miss something.

Reply:
ah, no, that would require more changes to the mdscache object. In any case, my point was that it's wasteful to add 4 bits per lock. very very wasteful.